File: 프로그래머스/2/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.py
from collections import deque,defaultdict
def solution(land):
    ans = 0
    directions = [(0,1),(1,0),(0,-1),(-1,0)]
    n = len(land)
    m = len(land[0])
    size = defaultdict(int)
    cnt = 1
    def bfs(x,y,cnt):
        s = 1
        q = deque()
        land[x][y] += cnt
        q.append((x,y))
        while q:
            cx,cy = q.popleft()
            for d in directions:
                nx,ny = cx+d[0],cy+d[1]
                if 0 <= nx < n and 0 <= ny < m and land[nx][ny] == 1:
                    land[nx][ny] += cnt
                    s += 1
                    q.append((nx,ny))
        size[land[x][y]] = s
        
    for i in range(n):
        for j in range(m):
            if land[i][j] == 1:
                bfs(i,j,cnt)
                cnt += 1
    # land[i][j] != 0 -> 시추 번호
    for j in range(m):
        visit = set()
        amount = 0
        # 열을 순회하면서 판단하기
        for i in range(n):
            # 이미 시추한 적이 있다면 pass
            if land[i][j] != 0 and land[i][j] not in visit:
                amount += size[land[i][j]]
                visit.add(land[i][j])
        ans = max(ans,amount)
    return ans

File: 프로그래머스/2/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.java
import java.util.*;
class Solution {
    static int[][] dir = {{0,1},{0,-1},{1,0},{-1,0}};
    static int n;
    static int m;
    static int cnt = 2;
    static Map<Integer, Integer> oil = new HashMap<>();
    public int solution(int[][] land) {
        int ans = 0;
        n = land.length;
        m = land[0].length;
        boolean[][] visit = new boolean[n][m];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (!visit[i][j] && land[i][j] == 1) {
                    bfs(i,j,visit,land);
                    cnt++;
                }
            }
        }
        // col을 순회하면서 시추 시작
        for (int c = 0; c < m; c++) {
            int count = 0;
            // 방문한 시추구역 번호 저장할 set
            Set<Integer> set = new HashSet<>();
            for (int r = 0; r < n; r++) {
                // 시추구역아니면 패스
                if (land[r][c] == 0) continue;
                int num = land[r][c];
                if (set.contains(num)) continue;
                set.add(num);
                count += oil.get(num);
            }
            ans = Math.max(ans,count);
        }
        return ans;
    }
    
    public static void bfs(int x, int y, boolean[][] visit, int[][] land) {
        Queue<int []> q = new ArrayDeque<>();
        visit[x][y] = true;
        q.add(new int[] {x,y});
        int total = 1;
        while(!q.isEmpty()) {
            int[] cur = q.poll();
            int cx = cur[0];
            int cy = cur[1];
            land[cx][cy] = cnt;
            for (int[] d : dir) {
                int nx = cx + d[0];
                int ny = cy + d[1];
                if (0 <= nx && nx < n && 0 <= ny && ny < m) {
                    if (!visit[nx][ny] && land[nx][ny] == 1) {
                        visit[nx][ny] = true;
                        q.add(new int[] {nx,ny});
                        total++;
                    }
                }
            }
        }
        // 다 순회하고 나서 map에 등록
        oil.put(cnt,total);
    }
}